### PR TITLE
feat: support tencent cloud SAML SSO authentication with casdoor 

### PR DESCRIPTION
**Pull request by Songjf-ttk**
Fix: https://github.com/casdoor/casdoor/issues/2168

### CODE REVIEW

**Comment by Songjf-ttk** [comment]
issue2168

File: web/src/table/SamlAttributeTable.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+  deleteRow(table, i) {
+    table = Setting.deleteRow(table, i);
+    this.updateTable(table);
+  }
+
+  renderTable(table) {
+    const columns = [
+      {
+        title: i18next.t("user:Name"),
+        dataIndex: "attributename",
```
Comment (hsluoyz) [inline web/src/table/SamlAttributeTable.js:56]: Does `attributeName` work? why not `name`?

File: web/src/table/SamlAttributeTable.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        render: (text, record, index) => {
+          return (
+            <Input value={text} onChange={e => {
+              this.updateField(table, index, "attributeName", e.target.value);
+            }} />
+          );
+        },
+      },
+      {
+        title: i18next.t("user:Name format"),
```
Comment (hsluoyz) [inline web/src/table/SamlAttributeTable.js:68]: nameformat -> nameFormat

File: web/src/table/SamlAttributeTable.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+          return (
+            <Input value={text} onChange={e => {
+              this.updateField(table, index, "value", e.target.value);
+            }} />
+          );
+        },
+      },
+      {
+        title: i18next.t("general:Action"),
+        dataIndex: "operation",
```
Comment (hsluoyz) [inline web/src/table/SamlAttributeTable.js:95]: action, not operation

File: object/application.go [hunk 4]
```diff
@@ -60,6 +66,7 @@ type Application struct {
 	CertPublicKey       string          `xorm:"-" json:"certPublicKey"`
 	Tags                []string        `xorm:"mediumtext" json:"tags"`
 	InvitationCodes     []string        `xorm:"varchar(200)" json:"invitationCodes"`
+	SamlAttribute       []*SamlItem     `xorm:"varchar(1000)" json:"samlAttribute"`
```
Comment (hsluoyz) [inline object/application.go:69]: SamlAttribute -> SamlAttributes

File: object/saml_idp.go [hunk 5]
```diff
@@ -103,9 +104,20 @@ func NewSamlResponse(user *User, host string, certificate string, destination st
 	displayName.CreateAttr("NameFormat", "urn:oasis:names:tc:SAML:2.0:attrname-format:basic")
 	displayName.CreateElement("saml:AttributeValue").CreateAttr("xsi:type", "xs:string").Element().SetText(user.DisplayName)
 
-	roles := attributes.CreateElement("saml:Attribute")
-	roles.CreateAttr("Name", "Roles")
-	roles.CreateAttr("NameFormat", "urn:oasis:names:tc:SAML:2.0:attrname-format:basic")
+	var roles *etree.Element
+	if strings.Contains(destination, "tencent") && application.SamlAttribute != nil {
```
Comment (hsluoyz) [inline object/saml_idp.go:108]: No need to use if

File: object/saml_idp.go [hunk 6]
```diff
@@ -103,9 +104,20 @@ func NewSamlResponse(user *User, host string, certificate string, destination st
...
+		for _, item := range application.SamlAttribute {
+			roles = attributes.CreateElement("saml:Attribute")
```
Comment (hsluoyz) [inline object/saml_idp.go:110]: roles or role? Isn't it only one role?

File: web/src/table/SamlAttributeTable.js [hunk 7]
```diff
@@ -0,0 +1,132 @@
+// Copyright 2021 The Casdoor Authors. All Rights Reserved.
```
Comment (hsluoyz) [inline web/src/table/SamlAttributeTable.js:1]: 2023

File: web/src/ApplicationEditPage.js [hunk 8]
```diff
@@ -104,6 +105,7 @@ class ApplicationEditPage extends React.Component {
       providers: [],
       uploading: false,
       mode: props.location.mode !== undefined ? props.location.mode : "edit",
+      samlAttribute: [],
```
Comment (hsluoyz) [inline web/src/ApplicationEditPage.js:108]: +s

File: object/saml_idp.go [hunk 9]
```diff
@@ -103,9 +104,20 @@ func NewSamlResponse(user *User, host string, certificate string, destination st
...
+			roles.CreateAttr("Name", item.AttributeName)
+			roles.CreateAttr("NameFormat", item.Nameformat)
+			roles.CreateAttr("FriendlyName", "tencent cloud")
+			roles.CreateElement("saml:AttributeValue").CreateAttr("xsi:type", "xs:string").Element().SetText(item.Value)
```
Comment (hsluoyz) [inline object/saml_idp.go:114]: Does `SamlAttributes` only used for SAML roles? If yes, why not rename to `SamlRoles`?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: object/saml_idp.go [hunk 10]
```diff
@@ -103,9 +104,20 @@ func NewSamlResponse(user *User, host string, certificate string, destination st
...
-	roles := attributes.CreateElement("saml:Attribute")
-	roles.CreateAttr("Name", "Roles")
-	roles.CreateAttr("NameFormat", "urn:oasis:names:tc:SAML:2.0:attrname-format:basic")
+	var roles *etree.Element
+	if strings.Contains(destination, "tencent") && application.SamlAttribute != nil {
+		for _, item := range application.SamlAttribute {
+			roles = attributes.CreateElement("saml:Attribute")
+			roles.CreateAttr("Name", item.AttributeName)
+			roles.CreateAttr("NameFormat", item.Nameformat)
+			roles.CreateAttr("FriendlyName", "tencent cloud")
```
Comment (hsluoyz) [inline object/saml_idp.go:113]: Configure `FriendlyName` in frontend too, remove all `Tencent cloud`

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: web/src/table/SamlAttributeTable.js (hunk 2 above)
Comment (Songjf-ttk) [reply web/src/table/SamlAttributeTable.js]: down

File: web/src/table/SamlAttributeTable.js (hunk 3 above)
Comment (Songjf-ttk) [reply web/src/table/SamlAttributeTable.js]: down

File: object/saml_idp.go (hunk 5 above)
Comment (Songjf-ttk) [reply object/saml_idp.go]: down

File: object/saml_idp.go (hunk 6 above)
Comment (Songjf-ttk) [reply object/saml_idp.go]: down

File: object/saml_idp.go (hunk 9 above)
Comment (Songjf-ttk) [reply object/saml_idp.go]: This attribute can include the user's name, email, role, department, identity, permissions, or any other information related to the user's identity.

File: web/src/ApplicationEditPage.js (hunk 8 above)
Comment (Songjf-ttk) [reply web/src/ApplicationEditPage.js]: down

File: object/saml_idp.go (hunk 10 above)
Comment (Songjf-ttk) [reply object/saml_idp.go]: down

File: web/src/table/SamlAttributeTable.js (hunk 7 above)
Comment (Songjf-ttk) [reply web/src/table/SamlAttributeTable.js]: down

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**